lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,259 @@
+                                                                       collection_meta['version']))
+
+    if os.path.exists(collection_output):
+        if os.path.isdir(collection_output):
+            raise AnsibleError("The output collection artifact %s already exists, "
+                               "but is a directory - aborting" % collection_output)
+        elif not force:
+            raise AnsibleError("The file %s already exists. You can use --force to re-create "
+                               "the collection artifact." % collection_output)
+
+    _build_collection_tar(collection_path, collection_output, collection_manifest, file_manifest)
+
+
+def _get_galaxy_yml(galaxy_yml_path):
+    mandatory_keys = frozenset(['namespace', 'name', 'version', 'authors'])
+    optional_strings = ('description', 'repository', 'documentation', 'homepage', 'issues')
+    optional_lists = ('license', 'tags', 'authors')  # authors isn't optional but this will ensure it is list
+    optional_dicts = ('dependencies',)
+    all_keys = frozenset(list(mandatory_keys) + list(optional_strings) + list(optional_lists) + list(optional_dicts) +
+                         ['repository', 'documentation', 'homepage_url', 'issues'])
+
+    try:
+        with open(to_bytes(galaxy_yml_path), 'rb') as g_yaml:
+            galaxy_yml = yaml.safe_load(g_yaml)
+    except YAMLError as err:
+        raise AnsibleError("Failed to parse the galaxy.yml at %s with the following error:\n%s"
+                           % (to_native(galaxy_yml_path), to_native(err)))
+
+    set_keys = set(galaxy_yml.keys())
+    missing_keys = mandatory_keys.difference(set_keys)
+    if len(missing_keys) > 0:
+        raise AnsibleError("The collection galaxy.yml at %s is missing the following mandatory keys: %s"
+                           % (to_native(galaxy_yml_path), ", ".join(missing_keys)))
+
+    extra_keys = set_keys.difference(all_keys)
+    if len(extra_keys) > 0:
+        display.warning("Found unknown keys in collection galaxy.yml at %s: %s"
+                        % (to_native(galaxy_yml_path), ", ".join(extra_keys)))
+
+    # Add the defaults if they have not been set
+    for optional_string in optional_strings:
+        if optional_string not in galaxy_yml:
+            galaxy_yml[optional_string] = ""
+
+    for optional_list in optional_lists:
+        list_val = galaxy_yml.get(optional_list, None)
+
+        if list_val is None:
+            galaxy_yml[optional_list] = []
+        elif not isinstance(list_val, list):
+            galaxy_yml[optional_list] = [list_val]
+
+    for optional_dict in optional_dicts:
+        if optional_dict not in galaxy_yml:
+            galaxy_yml[optional_dict] = {}
+
+    return galaxy_yml
+
+
+def _build_files_manifest(collection_path):
+    ignore_files = frozenset(['*.pyc', '*.retry'])
+    ignore_dirs = frozenset(['CVS', '.bzr', '.hg', '.git', '.svn', '__pycache__', '.tox'])
```

Comment:
Can this be generated from `.gitignore` contents?

Reply:
@webknjaz we have no gaurantees that the artifacts are being built from a git checkout.

Reply:
Also this is meant to cover other version control programs like subversion mercurial and so on. We do have https://docs.ansible.com/ansible/latest/reference_appendices/config.html#galaxy-role-skeleton-ignore and could create something similar for this if we wanted to.